### ISSUE TITLE
OSX preferences don't log as expected

### ISSUE THREAD

**Issue by boblord** [facebook/osquery#858]
I'm experimenting with osquery. While I've been able to get several items to log correctly (changing mount points, USB changes, and user changes), the basic OSX system preferences are not working for me. A few things get logged once, but it's a subset of all preferences.

I removed the DB per @marpaia's advice and that didn't change anything. 

The only preferences that seem to be logged are:
IOBluetoothUSBDFUTool
applessdstatistics
blued
com.apple.AOSNotification
com.apple.AOSNotification.notbackedup
com.apple.AppleMultitouchTrackpad
com.apple.CoreDuet
com.apple.CoreGraphics
com.apple.HIToolbox
com.apple.InputMethodKit.UserDictionary
com.apple.awdd.persistent
com.apple.driver.AppleBluetoothMultitouch.mouse
com.apple.driver.AppleBluetoothMultitouch.trackpad
com.apple.driver.AppleHIDMouse
com.apple.icloud.findmydeviced
com.apple.icloud.findmydeviced.notbackedup
com.apple.stackshot
com.apple.storeagent
com.apple.telemetry.battery.charge-cycle
com.apple.universalaccess
com.apple.wifi.keychain-format
com.apple.xpc.activity
com.apple.xpc.activity2
kCFPreferencesAnyApplication

Note that there are over 200 preferences when I use osqueryi which means the list above is just a fraction of the total number of prefs:
```
osqueryi 'select distinct domain from preferences order by domain'
```

Turning Bluetooth off generated logs, but not turning it back on.

I changed prefs for findmydeviced, and that did not generate logs.

When I run the server in --verbose mode I see these interesting warnings:
I0316 09:17:48.280386 108556288 scheduler.cpp:74] Executing query: select * from preferences;
I0316 09:17:48.287329 108556288 virtual_table.cpp:216] Table preferences row 0 did not include column subkey
I0316 09:17:48.287359 108556288 virtual_table.cpp:216] Table preferences row 0 did not include column path
I0316 09:17:48.287374 108556288 virtual_table.cpp:216] Table preferences row 1 did not include column value
I0316 09:17:48.287385 108556288 virtual_table.cpp:216] Table preferences row 1 did not include column path
I0316 09:17:48.287396 108556288 virtual_table.cpp:216] Table preferences row 2 did not include column value
I0316 09:17:48.287407 108556288 virtual_table.cpp:216] Table preferences row 2 did not include column path
I0316 09:17:48.287418 108556288 virtual_table.cpp:216] Table preferences row 3 did not include column value
... for 800+ rows. Am I hitting a timeout or other resource limit?

My actual goal at the moment is to see if the screensaver is turned on with an appropriate timeout:
```
osqueryi 'select * from preferences where domain = "com.apple.screensaver";'
```

I've also played with rules like this, but these didn't help either:
```json
{
        "name": "screensaver_prefs",
        "query": "select * from preferences where domain = 'com.apple.screensaver';",
        "interval": 10
}
```

Here is my osquery.conf file:
```json
{
    "options": {
        "config_plugin": "filesystem",
        "debug": "true",
        "event_pubsub_expiry": "86000",
        "host_identifier": "lord-mac",
        "log_result_events": "true",
        "logger_plugin": "filesystem",
        "schedule_splay_percent": "10",
        "verbose_debug": "true",
        "worker_threads": "4"
    },
    "scheduledQueries": [

    //does NOT generate expected prefs in the logs. Halp!
    {
        "interval": 15,
        "name": "changing_preferences",
        "query": "select * from preferences;"
    },

    //works as expected
    {
        "interval": 3600,
        "name": "osq_info",
        "query": "SELECT * FROM osquery_info;"
    },

    //works as expected
    {
        "interval": 3600,
        "name": "changing_processes",
        "query": "SELECT name, path, pid FROM processes;"
    },

    //works as expected
    {
        "interval": 30,
        "name": "change_mounts",
        "query": "SELECT device, device_alias, path, type,flags from mounts;"
    },

    //works as expected
    {
        "interval": 30,
        "name": "changing_USB",
        "query": "SELECT * FROM usb_devices;"
    },

    //works as expected
    {
        "interval": 60,
        "name": "changing_users",
        "query": "select * from users;"
    }
    ]
}
```

**Comment by theopolis** [comment]
Looking into this!

Just a note: `select * from preferences` is an expensive operation, it's the equivalent of `defaults read`. My guess is something in `*` is emitting A LOT of data, causing the daemon worker to allocate too much memory (~30M on my host), and it's then killed by the watchdog. You can relax the watchdog's instantaneous memory limits using `--watchdog_level=0`, sets the limit to 50M, or disable it completely `--disable_watchdog`.

Also note, `select * from preferences where domain = 'SOMETHING.HERE'` is much much less expensive. Expected output from "com.apple.screensaver":
```
>>> ./build/darwin/osquery/osqueryi "select * from preferences where domain = 'com.apple.screensaver';"

+-----------------------+---------------------+--------+-------+--------+------+
| domain                | key                 | subkey | value | forced | path |
+-----------------------+---------------------+--------+-------+--------+------+
| com.apple.screensaver | tokenRemovalAction  |        | 0     | 0      |      |
| com.apple.screensaver | askForPassword      |        | 1     | 0      |      |
| com.apple.screensaver | askForPasswordDelay |        | 0     | 0      |      |
+-----------------------+---------------------+--------+-------+--------+------+
```

**Comment by theopolis** [comment]
Hey @boblord was the `--disable_watchdog` flag helpful? I'm trying to find a way to limit the resources `preferences` needs. But that might take some time-- I want to make sure we unblock you ASAP. :)

**Comment by boblord** [comment]
Sadly it didn't change anything on the runs I tried. Is anyone else able to reproduce the issue? If not, I'll try another machine to see what I come up with. If anyone has a conf file that shows screensaver prefs I'll try that too. Thanks!

**Comment by theopolis** [comment]
`domain = 'com.apple.screensaver'` in OS X preferences (defaults) means read all preferences from `/Library/Preferences/*.plist`, + `~/Library/Preferences/*.plist`.

I think the problem here is when osqueryd is running as root, `~` does not include the user's preferences so `/Users/<YOU>/Library/Preferences/com.apple.screensaver.plist` is never read/reported. 

Test:
```
» osqueryi "select * from preferences where domain = 'com.apple.screensaver';"

+-----------------------+---------------------+--------+-------+--------+------+
| domain                | key                 | subkey | value | forced | path |
+-----------------------+---------------------+--------+-------+--------+------+
| com.apple.screensaver | tokenRemovalAction  |        | 0     | 0      |      |
| com.apple.screensaver | askForPassword      |        | 1     | 0      |      |
| com.apple.screensaver | askForPasswordDelay |        | 0     | 0      |      |
+-----------------------+---------------------+--------+-------+--------+------+
» sudo osqueryi "select * from preferences where domain = 'com.apple.screensaver';" 
»  
```

The solution is to set `where path = '/Users/<YOU>/Library/Preferences/com.apple.screensaver.plist'`:

```
» sudo osqueryi "select * from preferences where domain = 'com.apple.screensaver' and path = '/Users/reed/Library/Preferences/com.apple.screensaver.plist';"

+-----------------------+---------------------+--------+-------+--------+-------------------------------------------------------------+
| domain                | key                 | subkey | value | forced | path                                                        |
+-----------------------+---------------------+--------+-------+--------+-------------------------------------------------------------+
| com.apple.screensaver | askForPasswordDelay |        | 0     | 0      | /Users/reed/Library/Preferences/com.apple.screensaver.plist |
| com.apple.screensaver | askForPassword      |        | 1     | 0      | /Users/reed/Library/Preferences/com.apple.screensaver.plist |
| com.apple.screensaver | tokenRemovalAction  |        | 0     | 0      | /Users/reed/Library/Preferences/com.apple.screensaver.plist |
+-----------------------+---------------------+--------+-------+--------+-------------------------------------------------------------+
```

unfortunately we cannot represent a glob or wildcard operator for "YOU" right now. We're working towards that, we have globbing, but no way to tell SQLite the results are intended, and that the results will not match the equality operator given in the where clause. We'll soon be able to support something like `where path = '/Users/%/Library/Preferences/%'` using the SQLite MATCH operator and extended SQL. 

There's another alternative here, and that is us changing how `preferences` works, and enumerating all users then searching their preferences folder for plists by default. This will add a more latency though.

**Comment by boblord** [comment]
@theopolis: that syntax makes perfect sense and also works in my conf file. So thank you!

**Comment by theopolis** [comment]
I played with the OS X Preferences APIs again to try to limit the memory it allocated for keys/value retrieval with no luck. :( I also tried to grab user-specific preferences when running as root without having to set the explicit path. Failed on that too. :(

The best way to grab the domain-specific information for every user:
```sql
SELECT username, directory, p.*
FROM users, preferences p
WHERE users.directory || '/Library/Preferences/com.apple.screensaver.plist' = p.path ;
```

Let me know if you see anything weird or unexpected!
